### ISSUE TITLE
Type conversion rule fails after upgrading arrow-datafusion

### ISSUE THREAD

**Issue by v0y4g3r** [GreptimeTeam/greptimedb#1365]
### What type of bug is this?

Unexpected error

### What subsystems are affected?

Datanode

### What happened?

In GreptimeDB we rely on `TypeConversionRule` to enable operations between timestamp and other types like bigint and string literals. 

https://github.com/GreptimeTeam/greptimedb/blob/a3e47955b82ae1a77d6173640d577b3f62488aa9/src/query/src/optimizer.rs#L148-L150

But after [arrow-datafusion #5831](https://github.com/apache/arrow-datafusion/pull/5831) was merged last week, now type coercion happens before optimizer rewrites literal types, which will cause query failure.

### What operating system did you use?

macOS aarch64

### How can we reproduce the bug?

```bash

mysql> CREATE table demo1 (ts timestamp(3) TIME INDEX, cnt INT);
Query OK, 0 rows affected (0.07 sec)

mysql> select ts from demo1 where ts >= 0;
ERROR 1815 (HY000): Failed to execute query: select ts from demo1 where ts >= 0, source: Failed to execute logical plan, source: Failure during query execution, source: Fail to create physical plan: Error during planning: Timestamp(Millisecond, None) >= Int64 can't be evaluated because there isn't a common type to coerce the types to

```

**Comment by waynexia** [comment]
Https://github.com/apache/arrow-datafusion/pull/5963 might fix it

**Comment by killme2008** [comment]
Can we add a test case for this rule? We should find this bug when we upgrade the datafusion.

**Comment by waynexia** [comment]
It's found by existing tests when I try to upgrade datafusion

**Comment by jacksonrnewhouse** [comment]
Over at https://github.com/ArroyoSystems/arroyo we also ran into this issue, as we were dependent on the optimizer performing the analysis. Fixable by also running the default analyzer separately, but definitely a surprising change in behavior.

**Comment by waynexia** [comment]
> Over at https://github.com/ArroyoSystems/arroyo we also ran into this issue, as we were dependent on the optimizer performing the analysis. Fixable by also running the default analyzer separately, but definitely a surprising change in behavior.

Thanks for that information. This is indeed surprising and it takes me half day to pin it down. We've fixed it in #1386 by [changing our rule into an analyzer rule](https://github.com/GreptimeTeam/greptimedb/pull/1386/files#diff-0c63dac4865882b65621f642682be1c18953486daf6a27bf2ca12f8788961997) and make sure it's the first rule to be applied. Hope this works in your situation.

**Comment by waynexia** [comment]
Close as completed. Feel free to continue the discussion.

**Comment by jacksonrnewhouse** [comment]
Yeah, we solved it similarly with https://github.com/ArroyoSystems/arroyo/pull/92

> > Over at https://github.com/ArroyoSystems/arroyo we also ran into this issue, as we were dependent on the optimizer performing the analysis. Fixable by also running the default analyzer separately, but definitely a surprising change in behavior.
> 
> Thanks for that information. This is indeed surprising and it takes me half day to pin it down. We've fixed it in #1386 by [changing our rule into an analyzer rule](https://github.com/GreptimeTeam/greptimedb/pull/1386/files#diff-0c63dac4865882b65621f642682be1c18953486daf6a27bf2ca12f8788961997) and make sure it's the first rule to be applied. Hope this works in your situation.

Yeah, we solved
